### PR TITLE
Node monitor (extended subscription)

### PR DESCRIPTION
**Pull request by kung-foo**
This PR creates a new `NodeMonitor` struct for creating managed subscriptions. It attempts to hide implementation details from the user, and provide a simple, high-level interface for interacting with data change events.
 
see related discussion: https://github.com/gopcua/opcua/issues/240

The example code shows the two different methods for a subscription: callback-based, and channel-based.

### CODE REVIEW

**Comment by kung-foo** [comment]
There are some naming/structure issues I'd still like to resolve. Possibly merge `NodeMonitor` and `Subscription` into a single struct and move `Subscribe(...)` and `ChanSubscribe(...)` to package level functions (both returning the new, unified struct).

**Comment by magiconair** [comment]
This looks like a good start. Maybe call it `monitor.Subscription` instead of `monitor.New`. We might have other monitors. I'll need something that reads a list of nodes on a regular basis and that would fit in this quite nicely (e.g. `monitor.Nodes`).

Do we have to support both channel and callback mechanisms? Wouldn't channel be enough? 

Also, you can add nodes as strings and `NodeIDs` but when you set up the monitor you can use only strings. Don't have a good idea on how to solve that right now but might be something to keep in mind. With Channel and Callback you potentially get four different combinations. 

You probably don't want to `panic` when the subscription id doesn't match ... :)

**Comment by kung-foo** [comment]
Thanks for the feedback. Regarding chan v callback, my favorite APIs are the ones that usually provide both. From this somewhat hyperbolic article: https://www.jtolio.com/2016/03/go-channels-are-bad-and-you-should-feel-bad/
> [...] your API can always be more general, always more flexible, and take drastically less resources if you use callbacks instead of channels.

**Comment by kung-foo** [comment]
I should point out that the _current_ callback implementation is actually _less_ performant the the channel method because the callback code uses a channel behind the scenes. However, that could be removed with a little bit of work.

File: subscription.go [hunk 1]
```diff
@@ -196,6 +198,29 @@ func (s *Subscription) sendNotification(ctx context.Context, data *PublishNotifi
 
 }
 
+// Stats returns a diagnostic struct with metadata about the current subscription
+func (s *Subscription) Stats() (*ua.SubscriptionDiagnosticsDataType, error) {
+	// TODO(kung-foo): once browsing feature is merged, attempt to get direct access to the
```
Comment (magiconair) [inline subscription.go:203]: does this refer to the more browsing branch or something else?

**Comment by magiconair** [comment]
@kung-foo are you using this code right now or are you planning on using it?

File: subscription.go (hunk 1 above)
Comment (kung-foo) [reply subscription.go:203]: yes. though I haven't tried it on this exact use case. So it is possible that I can't get access to the single diag node. Also, I noticed that with the FreeOPCUA server, no diag nodes are generated. Is there a standard way to indicate that that feature is not supported?

**Comment by kung-foo** [comment]
Yes, we are using it, though not in production.

File: subscription.go (hunk 1 above)
Comment (magiconair) [reply subscription.go:203]: I don't know.

**Comment by magiconair** [comment]
How did you find the timer leak?

**Comment by kung-foo** [comment]
re the timer leak: manual inspection. I just happen to remember that "idiosyncrasy" of the timer API.

**Comment by magiconair** [comment]
@kung-foo How far are we from merging this? I'd like to start using it.

**Comment by kung-foo** [comment]
I think we are good to merge. We can re-visit features like reconnecting to a subscription in a future PR.

**Comment by magiconair** [comment]
Then lets do it.

**Comment by fdamador** [comment]
Thank you for creating the monitor code which i've been testing.  

I have a need to pass the msg.value.value() values data back to my own custom struct array I have in the main() routine for a series of NodeIDs.  Can I do this by interrogating the subscription server?  Or do I have to send the NodeIds data (time, value, quality, etc) via channel?

**Pull request merged**